src/trie/merkle_tree.rs

```diff
@@ -704,9 +740,162 @@ impl<H: HasherT, DB: BonsaiDatabase, ID: Id> MerkleTree<H, DB, ID> {
         self.db.contains(&TrieKeyType::Flat(key.to_vec()))
     }
 
-    /// preload_nodes from the current root towards the destination [Leaf](Node::Leaf) node.
-    /// Returns the list of nodes along the path.
+    /// Generates a merkle-proof for a given `key`.
+    ///
+    /// if it exists, or down to the node which proves that the key does not exist.
+    ///
+    /// The nodes are returned in order, root first.
+    ///
+    /// Verification is performed by confirming that:
+    ///   1. the chain follows the path of `key`, and
+    ///   2. the hashes are correct, and
+    ///   3. the root hash matches the known root
     ///
+    /// # Arguments
+    ///
+    /// * `key` - The key to get the merkle proof of.
+    ///
+    /// # Returns
+    ///
+    /// The merkle proof and all the child nodes hashes.
+    pub fn get_proof(&self, key: &BitSlice<u8, Msb0>) -> Result<Vec<ProofNode>, BonsaiStorageError>
```

Comment:
can we make that generic over the hash function ?

Reply:
Can you point me to the code that is "hash-specific" in this function ?

Reply:
ah indeed all good here